docarray/array/storage/redis/backend.py

```diff
@@ -0,0 +1,171 @@
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Any, Dict, Iterable, List, Optional, Tuple, Union
+
+import numpy as np
+from docarray import Document
+from docarray.array.storage.base.backend import BaseBackendMixin, TypeMap
+from docarray.helper import dataclass_from_dict
+
+from redis import Redis
+from redis.commands.search.field import NumericField, TextField, VectorField
+
+if TYPE_CHECKING:
+    from ....typing import ArrayType, DocumentArraySourceType
+
+
+@dataclass
+class RedisConfig:
+    n_dim: int
+    host: str = field(default='localhost')
+    port: int = field(default=6379)
+    flush: bool = field(default=False)
+    update_schema: bool = field(default=True)
+    distance: str = field(default='COSINE')
+    redis_config: Dict[str, Any] = field(default_factory=dict)
+    index_text: bool = field(default=False)
+    tag_indices: List[str] = field(default_factory=list)
+    batch_size: int = field(default=64)
+    method: str = field(default='HNSW')
+    initial_cap: Optional[int] = None
+    ef_construction: Optional[int] = None
+    m: Optional[int] = None
+    ef_runtime: Optional[int] = None
+    block_size: Optional[int] = None
+    columns: Optional[List[Tuple[str, str]]] = None
+
+
+class BackendMixin(BaseBackendMixin):
+    """Provide necessary functions to enable this storage backend."""
+
+    TYPE_MAP = {
+        'str': TypeMap(type='text', converter=TextField),
+        'bytes': TypeMap(type='text', converter=TextField),
+        'int': TypeMap(type='integer', converter=NumericField),
+        'float': TypeMap(type='float', converter=NumericField),
+        'double': TypeMap(type='double', converter=NumericField),
+        'long': TypeMap(type='long', converter=NumericField),
+        # TODO add bool
+    }
+
+    def _init_storage(
+        self,
+        _docs: Optional['DocumentArraySourceType'] = None,
+        config: Optional[Union[RedisConfig, Dict]] = None,
+        **kwargs,
+    ):
+        if not config:
+            raise ValueError('Empty config is not allowed for Redis storage')
+        elif isinstance(config, dict):
+            config = dataclass_from_dict(RedisConfig, config)
+
+        if config.distance not in ['L2', 'IP', 'COSINE']:
+            raise ValueError(f'Distance metric {config.distance} not supported')
```

Comment:
something like this?
f'Expecting distance metric one of COSINE, L2 OR IP, got {config.distance} instead'